arangod/Aql/GraphOptimizerRules.cpp

```diff
@@ -0,0 +1,206 @@
+////////////////////////////////////////////////////////////////////////////////
+/// DISCLAIMER
+///
+/// Copyright 2023-2023 ArangoDB GmbH, Cologne, Germany
+///
+/// Licensed under the Apache License, Version 2.0 (the "License");
+/// you may not use this file except in compliance with the License.
+/// You may obtain a copy of the License at
+///
+///     http://www.apache.org/licenses/LICENSE-2.0
+///
+/// Unless required by applicable law or agreed to in writing, software
+/// distributed under the License is distributed on an "AS IS" BASIS,
+/// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+/// See the License for the specific language governing permissions and
+/// limitations under the License.
+///
+/// Copyright holder is ArangoDB GmbH, Cologne, Germany
+///
+/// @author Michael Hackstein
+////////////////////////////////////////////////////////////////////////////////
+
+#include "GraphOptimizerRules.h"
+
+#include "Aql/ExecutionNode.h"
+#include "Aql/ExecutionPlan.h"
+#include "Aql/Optimizer.h"
+#include "Aql/TraversalNode.h"
+#include "Containers/SmallVector.h"
+
+#include "Logger/LogMacros.h"
+
+using namespace arangodb;
+using namespace arangodb::aql;
+
+namespace {
+
+enum class PathAccessState {
+  // Search for access to any path variable
+  SEARCH_ACCESS_PATH,
+  // Check if we access edges or vertices on it
+  ACCESS_EDGES_OR_VERTICES,
+  // Check if we have an indexed access [x] (we find the x here)
+  SPECIFIC_DEPTH_ACCESS_DEPTH,
+  // CHeck if we have an indexed access [x] (we find [ ] here)
+  SPECIFIC_DEPTH_ACCESS_INDEXED_ACCESS,
```

Comment:
According to our naming conventions, these should be named `kSearchAccessPath`, `kAccessEdgesOrVertices`, ...